lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -0,0 +1,1288 @@
+  return unknownResult();
+}
+
+// Handle calls to opaque callees, either by handling them and returning None or
+// by returning with a Unknown indicating a failure.
+llvm::Optional<SymbolicValue>
+ConstExprFunctionState::computeOpaqueCallResult(ApplyInst *apply,
+                                                SILFunction *callee) {
+  LLVM_DEBUG(llvm::dbgs() << "ConstExpr Opaque Callee: " << *callee << "\n");
+  return evaluator.getUnknown((SILInstruction *)apply, UnknownReason::Default);
+}
+
+// TODO: Refactor this to someplace common, this is defined in Devirtualize.cpp.
+SubstitutionMap getWitnessMethodSubstitutions(SILModule &Module, ApplySite AI,
+                                              SILFunction *F,
+                                              ProtocolConformanceRef CRef);
+
+/// Given a call to a function, determine whether it is a call to a constexpr
+/// function.  If so, collect its arguments as constants, fold it and return
+/// None.  If not, mark the results as Unknown, and return an Unknown with
+/// information about the error.
+llvm::Optional<SymbolicValue>
+ConstExprFunctionState::computeCallResult(ApplyInst *apply) {
+  auto conventions = apply->getSubstCalleeConv();
+
+  // Determine the callee.
+  auto calleeFn = getConstantValue(apply->getOperand(0));
+  if (calleeFn.getKind() != SymbolicValue::Function)
+    return evaluator.getUnknown((SILInstruction *)apply,
+                                UnknownReason::Default);
+
+  SILFunction *callee = calleeFn.getFunctionValue();
+
+  // Verify that we can fold all of the arguments to the call.
+  SmallVector<SymbolicValue, 4> paramConstants;
+  for (unsigned i = 0, e = apply->getNumOperands() - 1; i != e; ++i) {
+    // If any of the arguments is a non-constant value, then we can't fold this
+    // call.
+    auto op = apply->getOperand(i + 1);
+    SymbolicValue argValue = getConstantValue(op);
+    if (!argValue.isConstant())
+      return argValue;
+    paramConstants.push_back(argValue);
+  }
+
+  // If we reached an external function that hasn't been deserialized yet, make
+  // sure to pull it in so we can see its body.  If that fails, then we can't
+  // analyze the function.
+  if (callee->isExternalDeclaration()) {
+    callee->getModule().loadFunction(callee);
+    if (callee->isExternalDeclaration())
+      return computeOpaqueCallResult(apply, callee);
+  }
+
+  // Compute the substitution map for the callee, which maps from all of its
+  // generic requirements to concrete conformances and concrete types.
+  SubstitutionMap calleeSubMap;
+
+  auto calleeFnType = callee->getLoweredFunctionType();
+  if (calleeFnType->getGenericSignature()) {
+    ApplySite AI(apply);
+
+    // Get the substitution map of the call.  This maps from the callee's space
+    // into the caller's world.
+    SubstitutionMap callSubMap;
+    if (calleeFnType->getRepresentation() ==
+        SILFunctionType::Representation::WitnessMethod) {
```

Comment:
If you have a class method, you also need to do some remapping for vtable substitutions. Eg if I have `class G<T, U> : C<(T, U), Int>`, and I'm calling a method on an instance of G<String, Float>, but its actually implemented on C, I have to remap the substitution map to `C<(String, Float), Int>`.

Reply:
Should we assert here that the function is not a class method. Is there a way to do that?

Reply:
Looks like this may work: 
`assert(!calleeFnType->getSelfInstanceType()->getClassOrBoundGenericClass(), "class methods are not supported")`

Reply:
Perhaps this is better:
`assert(!calleeFnType->hasSelfParam() || !calleeFnType->getSelfInstanceType()->getClassOrBoundGenericClass())`